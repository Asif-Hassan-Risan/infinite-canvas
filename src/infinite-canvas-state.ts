import { Transformation } from "./transformation";
import { DrawingInstruction } from "./drawing-instruction";

export class InfiniteCanvasState implements DrawingInstruction{
    constructor(
        public readonly lineWidth: number,
        public readonly lineDashOffset: number,
        public readonly lineDash: number[],
        public readonly fillStyle: string | CanvasGradient | CanvasPattern,
        public readonly strokeStyle: string | CanvasGradient | CanvasPattern
    ){

    }
    public apply(context: CanvasRenderingContext2D, transformation: Transformation): void{
        context.lineWidth = this.lineWidth * transformation.scale;
        context.lineDashOffset = this.lineDashOffset * transformation.scale;
        context.setLineDash(this.lineDash.map(s => s * transformation.scale));
        context.fillStyle = this.fillStyle;
        context.strokeStyle = this.strokeStyle;
    }
    public setLineWidth(lineWidth: number): InfiniteCanvasState{
        return new InfiniteCanvasState(
            lineWidth,
            this.lineDashOffset,
            this.lineDash,
            this.fillStyle,
            this.strokeStyle
        );
    }
    public setLineDashOffset(lineDashOffset: number): InfiniteCanvasState{
        return new InfiniteCanvasState(
            this.lineWidth,
            lineDashOffset,
            this.lineDash,
            this.fillStyle,
            this.strokeStyle
        );
    }
    public setLineDash(lineDash: number[]): InfiniteCanvasState{
        return new InfiniteCanvasState(
            this.lineWidth,
            this.lineDashOffset,
            lineDash,
            this.fillStyle,
            this.strokeStyle
        );
    }
    public setFillStyle(fillStyle: string | CanvasGradient | CanvasPattern): InfiniteCanvasState{
        return new InfiniteCanvasState(
            this.lineWidth,
            this.lineDashOffset,
            this.lineDash,
            fillStyle,
            this.strokeStyle
        );
    }
    public setStrokeStyle(strokeStyle: string | CanvasGradient | CanvasPattern): InfiniteCanvasState{
        return new InfiniteCanvasState(
            this.lineWidth,
            this.lineDashOffset,
            this.lineDash,
            this.fillStyle,
            strokeStyle
        );
    }
    public static default(): InfiniteCanvasState{
        return new InfiniteCanvasState(
            1,
            0,
            [],
            "#000",
            "#000"
        );
    }
}