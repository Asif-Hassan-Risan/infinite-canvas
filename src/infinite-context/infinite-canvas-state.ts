import { ViewBox } from "../viewbox";

export class InfiniteCanvasState implements CanvasState{
	constructor(private viewBox: ViewBox){}
	public restore():void{
		this.viewBox.restoreState();
	}
	public save(): void{
		this.viewBox.saveState();
	}
}