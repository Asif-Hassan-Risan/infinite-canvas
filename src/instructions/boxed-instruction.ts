import { Instruction } from "./instruction";

export interface BoxedInstruction extends Instruction{
    replaceInnerInstructionWith(instruction: Instruction): void;
    copy(): BoxedInstruction;
}