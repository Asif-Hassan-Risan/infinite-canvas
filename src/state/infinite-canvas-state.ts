import { InfiniteCanvasStateInstance } from "./infinite-canvas-state-instance";
import { StateChange } from "./state-change";
import { ChainableStateChange } from "./chainable-state-change";

export class InfiniteCanvasState{
    constructor(public current: InfiniteCanvasStateInstance, public stack: InfiniteCanvasStateInstance[] = []){}
    public static default: InfiniteCanvasState = new InfiniteCanvasState(InfiniteCanvasStateInstance.default, []);
    public withChangedState(change: (state: InfiniteCanvasStateInstance) => StateChange<InfiniteCanvasStateInstance>): StateChange<InfiniteCanvasState>{
        const stateChange: StateChange<InfiniteCanvasStateInstance> = change(this.current);
        return {
            newState: new InfiniteCanvasState(stateChange.newState, this.stack),
            instruction: stateChange.instruction
        };
    }
    public save(): StateChange<InfiniteCanvasState>{
        return {
            newState: new InfiniteCanvasState(this.current, (this.stack || []).concat([this.current])),
            instruction: (context: CanvasRenderingContext2D) => {context.save();}
        };
    }
    public restore(): StateChange<InfiniteCanvasState>{
        if(!this.stack || this.stack.length === 0){
            return {
                newState: this,
                instruction: undefined
            };
        }
        const topOfStack: InfiniteCanvasStateInstance = this.stack[this.stack.length - 1];
        return {
            newState: new InfiniteCanvasState(topOfStack, this.stack.slice(0, this.stack.length - 1)),
            instruction: (context: CanvasRenderingContext2D) => {context.restore();}
        };
    }
    public convertTo(other: InfiniteCanvasState): StateChange<InfiniteCanvasState>{
        let change: ChainableStateChange<InfiniteCanvasState> = new ChainableStateChange(this, []);
        const indexOfHighestCommon: number = InfiniteCanvasState.findIndexOfHighestCommon(this.stack, other.stack);
        let currentIndex: number = indexOfHighestCommon + 1;
        for(let i: number = this.stack.length - 1; i > indexOfHighestCommon; i--){
            change = change.change(s => s.restore());
        }
        while(currentIndex < other.stack.length){
            change = change.change(s => s.withChangedState(ss => ss.convertToState(other.stack[currentIndex])));
            change = change.change(s => s.save());
            currentIndex++;
        }
        change = change.change(s => s.withChangedState(ss => ss.convertToState(other.current)));
        return change;
    }

    private static findIndexOfHighestCommon(one: InfiniteCanvasStateInstance[], other: InfiniteCanvasStateInstance[]): number{
        let result: number = 0;
        let lowestIndex: number = Math.min(one.length - 1, other.length - 1);
        while(result <= lowestIndex){
            if(!one[result].equals(other[result])){
                return result - 1;
            }
            result++;
        }
        return result - 1;
    }
}