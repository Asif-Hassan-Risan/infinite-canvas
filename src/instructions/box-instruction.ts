import { Instruction } from "./instruction";
import { BoxedInstruction } from "./boxed-instruction";
import { Transformation } from "../transformation";

export function boxInstruction(instruction: Instruction): BoxedInstruction{
    const result: BoxedInstruction = <BoxedInstruction>((context: CanvasRenderingContext2D, transformation: Transformation) => {
        instruction(context, transformation);
    });
    result.replaceInnerInstructionWith = (newInstruction: Instruction) => instruction = newInstruction;
    result.copy = () => boxInstruction(instruction);
    return result;
}