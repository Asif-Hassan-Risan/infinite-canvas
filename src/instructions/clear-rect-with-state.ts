import { StateAndInstruction } from "./state-and-instruction";
import { StateChangingInstructionSetWithArea } from "../interfaces/state-changing-instruction-set-with-area";
import { InfiniteCanvasState } from "../state/infinite-canvas-state";
import { Instruction } from "./instruction";
import { Area } from "../areas/area";
import { Rectangle } from "../areas/rectangle";

export class ClearRectWithState extends StateAndInstruction implements StateChangingInstructionSetWithArea{
    constructor(initialState: InfiniteCanvasState, state: InfiniteCanvasState, instruction: Instruction, combinedInstruction: Instruction, public area: Rectangle){
        super(initialState, state, instruction, combinedInstruction);
    }
    public hasDrawingAcrossBorderOf(area: Area): boolean{
        return false;
    }
    public intersects(area: Area): boolean{
        return this.area.intersects(area);
    }
    public isContainedBy(area: Rectangle): boolean {
        return area.contains(this.area);
    }
    public static createClearRect(initialState: InfiniteCanvasState, area: Rectangle, instructionToClear: Instruction): ClearRectWithState{
        return new ClearRectWithState(initialState, initialState, instructionToClear, instructionToClear, area);
    }
}