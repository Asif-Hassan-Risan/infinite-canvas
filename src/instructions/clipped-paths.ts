import { StateChangingInstructionSet } from "../interfaces/state-changing-instruction-set";
import { StateChangingInstructionSetWithAreaAndCurrentPath } from "../interfaces/state-changing-instruction-set-with-area-and-current-path";
import { Area } from "../areas/area";
import { InfiniteCanvasState } from "../state/infinite-canvas-state";
import { Instruction } from "./instruction";
import { Transformation } from "../transformation";
import { instructionSequence } from "../instruction-utils";

export class ClippedPaths {
    constructor(public area: Area, public latestClippedPath: StateChangingInstructionSet, public readonly previouslyClippedPaths?: ClippedPaths){}
    public withClippedPath(latestClippedPath: StateChangingInstructionSetWithAreaAndCurrentPath): ClippedPaths{
        const newArea: Area = latestClippedPath.getClippedArea(this.area);
        return new ClippedPaths(newArea, latestClippedPath, this);
    }
    public get initialState(): InfiniteCanvasState{
        return this.previouslyClippedPaths ? this.previouslyClippedPaths.initialState : this.latestClippedPath.initialState;
    }
    public except(other: ClippedPaths): ClippedPaths{
        if(other === this){
            return undefined;
        }
        if(this.previouslyClippedPaths){
            return new ClippedPaths(this.area, this.latestClippedPath, this.previouslyClippedPaths.except(other))
        }
        return this;
    }
    public contains(other: ClippedPaths): boolean{
        if(!other){
            return false;
        }
        if(this === other){
            return true;
        }
        if(this.previouslyClippedPaths){
            return this.previouslyClippedPaths.contains(other);
        }
        return false;
    }
    public getInstructionToRecreate(): Instruction{
        const instructionToRecreateLatest: Instruction = (context: CanvasRenderingContext2D, transformation: Transformation) => {
            this.latestClippedPath.execute(context, transformation);
        };
        if(this.previouslyClippedPaths){
            const instructionToRecreatePrevious: Instruction = this.previouslyClippedPaths.getInstructionToRecreate();
            const instructionToConvertToLatest: Instruction = this.previouslyClippedPaths.latestClippedPath.state.getInstructionToConvertToState(this.latestClippedPath.initialState);
            return instructionSequence(instructionToRecreatePrevious, instructionToConvertToLatest, instructionToRecreateLatest);
        }
        return instructionToRecreateLatest;
    }
}
