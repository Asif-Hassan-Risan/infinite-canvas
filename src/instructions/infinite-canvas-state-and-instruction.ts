import { Instruction } from "./instruction";
import { InfiniteCanvasState } from "../state/infinite-canvas-state";
import { StateAndInstruction } from "./state-and-instruction";

export class InfiniteCanvasStateAndInstruction extends StateAndInstruction{
    constructor(initialState: InfiniteCanvasState, initialInstruction: Instruction, stateChangeInstruction: Instruction, currentState: InfiniteCanvasState, initialStateChangeInstruction: Instruction, stateForInstruction: InfiniteCanvasState){
        super(initialState, initialInstruction, stateChangeInstruction, currentState, initialStateChangeInstruction, stateForInstruction);
    }
    public copy(): InfiniteCanvasStateAndInstruction{
        return new InfiniteCanvasStateAndInstruction(this.initialState, this.initialInstruction, this.stateChangeInstruction, this.state, this.initialStateChangeInstruction, this.stateForInstruction);
    }
    public static create(initialState: InfiniteCanvasState, initialInstruction: Instruction): InfiniteCanvasStateAndInstruction{
        return new InfiniteCanvasStateAndInstruction(initialState, initialInstruction, undefined, initialState, undefined, initialState);
    }
}