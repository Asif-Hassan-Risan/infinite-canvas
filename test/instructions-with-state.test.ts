import { InfiniteCanvasStateAndInstruction } from "../src/instructions/infinite-canvas-state-and-instruction";
import { InfiniteCanvasState } from "../src/state/infinite-canvas-state";
import { logWithState } from "./log-with-state";
import { defaultState } from "../src/state/default-state";

describe("a set of intructions that is only about state", () => {
    let instructionsWithState: InfiniteCanvasStateAndInstruction;

    beforeEach(() => {
        instructionsWithState = new InfiniteCanvasStateAndInstruction(defaultState, () => {});
    });

    describe("that receives a change", () => {

        beforeEach(() => {
            instructionsWithState.changeState(s => s.setFillStyle("#f00"));
        });

        it("should have recorded the change", () => {
            expect(logWithState(instructionsWithState)).toMatchSnapshot();
        });

        describe("and then receives a change that reverts that one", () => {

            beforeEach(() => {
                instructionsWithState.changeState(s => s.setFillStyle("#000"));
            });

            it("should contain no instructions anymore", () => {
                expect(logWithState(instructionsWithState)).toMatchSnapshot();
            });
        });

        describe("and then receives a change on a different property", () => {

            beforeEach(() => {
                instructionsWithState.changeState(s => s.setStrokeStyle("#ff0"));
            });

            it("should now contain two instructions", () => {
                expect(logWithState(instructionsWithState)).toMatchSnapshot();
            });
        });

        describe("and then receives a change on the same property", () => {

            beforeEach(() => {
                instructionsWithState.changeState(s => s.setFillStyle("#00f"));
            });

            it("should still contain only one instruction", () => {
                expect(logWithState(instructionsWithState)).toMatchSnapshot();
            });
        });
    });
});