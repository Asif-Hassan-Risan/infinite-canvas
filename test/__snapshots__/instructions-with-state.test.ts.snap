// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a set of intructions that is only about state that receives a change and then receives a change on a different property should now contain two instructions 1`] = `
Array [
  "context.fillStyle = \\"#f00\\"",
  "context.strokeStyle = \\"#ff0\\"",
]
`;

exports[`a set of intructions that is only about state that receives a change and then receives a change on the same property should still contain only one instruction 1`] = `
Array [
  "context.fillStyle = \\"#00f\\"",
]
`;

exports[`a set of intructions that is only about state that receives a change and then receives a change that reverts that one should contain no instructions anymore 1`] = `Array []`;

exports[`a set of intructions that is only about state that receives a change should have recorded the change 1`] = `
Array [
  "context.fillStyle = \\"#f00\\"",
]
`;
